File: leetcode/linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None
        i = 0
        node = head
        visited = {}
        while node.next:
            if node in visited.keys():
                return node
            visited[node] = i
            i += 1
            node = node.next
        return None

File: leetcode/linked-list-cycle-ii.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function detectCycle(head: ListNode | null): ListNode | null {
    let visited = new Set();
    let curr = head;
    while (curr != null) {
        if (visited.has(curr)) {
            return curr;
        }
        visited.add(curr);
        curr = curr.next;
    }
    return null;
};